### PR TITLE
fix: handle when FileSystem already exists

### LINKED ISSUE #1138: [BUG] failed to create driver when running from a jar artifact
**Context:**
- Playwright Version: 1.28.0
- Operating System: Arch Linux
- Browser: Chrominum
- Extra: -

## System:
 - OS: Linux 6.0 Arch Linux
 - Memory: 7.30 GB / 15.48 GB
 - Container: Yes
## Binaries:
 - Node: 14.21.0 - /usr/bin/node
 - Yarn: 1.22.19 - /usr/bin/yarn
 - npm: 8.19.2 - /usr/bin/npm
## Languages:
 - Bash: 5.1.16 - /usr/bin/bash

My code runs fine when running using gradle run. The issue happens when I build an uber jar using `shadowJar` and run the app from the jar file. 

The code is available at https://github.com/Suwayomi/Tachidesk-Server/blob/f47dc6b9dea927e55179186d07afac1ad9927be4/server/src/main/kotlin/eu/kanade/tachiyomi/network/interceptor/CloudflareInterceptor.kt

Here's the stack trace:

```
java.io.IOException: java.lang.RuntimeException: Failed to create driver
	at eu.kanade.tachiyomi.network.interceptor.CloudflareInterceptor.intercept(CloudflareInterceptor.kt:52)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109)
	at okhttp3.logging.HttpLoggingInterceptor.intercept(HttpLoggingInterceptor.kt:221)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109)
	at eu.kanade.tachiyomi.network.interceptor.UserAgentInterceptor.intercept(UserAgentInterceptor.kt:19)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109)
	at okhttp3.internal.connection.RealCall.getResponseWithInterceptorChain$okhttp(RealCall.kt:201)
	at okhttp3.internal.connection.RealCall.execute(RealCall.kt:154)
	at eu.kanade.tachiyomi.network.OkHttpExtensionsKt$asObservable$1$requestArbiter$1.request(OkHttpExtensions.kt:32)
	at rx.Subscriber.setProducer(Subscriber.java:211)
	at rx.internal.operators.OnSubscribeMap$MapSubscriber.setProducer(OnSubscribeMap.java:102)
	at rx.Subscriber.setProducer(Subscriber.java:205)
	at eu.kanade.tachiyomi.network.OkHttpExtensionsKt.asObservable$lambda$0(OkHttpExtensions.kt:54)
	at rx.Observable.unsafeSubscribe(Observable.java:10327)
	at rx.internal.operators.OnSubscribeDoOnEach.call(OnSubscribeDoOnEach.java:41)
	at rx.internal.operators.OnSubscribeDoOnEach.call(OnSubscribeDoOnEach.java:30)
	at rx.Observable.unsafeSubscribe(Observable.java:10327)
	at rx.internal.operators.OnSubscribeMap.call(OnSubscribeMap.java:48)
	at rx.internal.operators.OnSubscribeMap.call(OnSubscribeMap.java:33)
	at rx.internal.operators.OnSubscribeLift.call(OnSubscribeLift.java:48)
	at rx.internal.operators.OnSubscribeLift.call(OnSubscribeLift.java:30)
	at rx.Observable.subscribe(Observable.java:10423)
	at rx.Observable.subscribe(Observable.java:10390)
	at suwayomi.tachidesk.manga.impl.util.lang.RxCoroutineBridgeKt.awaitOne(RxCoroutineBridge.kt:29)
	at suwayomi.tachidesk.manga.impl.util.lang.RxCoroutineBridgeKt.awaitSingle(RxCoroutineBridge.kt:24)
	at suwayomi.tachidesk.manga.impl.MangaList.getMangaList(MangaList.kt:35)
	at suwayomi.tachidesk.manga.controller.SourceController$popular$2$1.invokeSuspend(SourceController.kt:78)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106)
	at kotlinx.coroutines.internal.LimitedDispatcher.run(LimitedDispatcher.kt:42)
	at kotlinx.coroutines.scheduling.TaskImpl.run(Tasks.kt:95)
	at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:570)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:750)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:677)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:664)
Caused by: java.lang.RuntimeException: Failed to create driver
	at com.microsoft.playwright.impl.driver.Driver.createAndInstall(Driver.java:105)
	at com.microsoft.playwright.impl.driver.Driver.ensureDriverInstalled(Driver.java:57)
	at com.microsoft.playwright.impl.PlaywrightImpl.createImpl(PlaywrightImpl.java:45)
	at com.microsoft.playwright.impl.PlaywrightImpl.create(PlaywrightImpl.java:35)
	at com.microsoft.playwright.Playwright.create(Playwright.java:96)
	at com.microsoft.playwright.Playwright.create(Playwright.java:100)
	at eu.kanade.tachiyomi.network.interceptor.CFClearance.resolveWithWebView(CloudflareInterceptor.kt:76)
	at eu.kanade.tachiyomi.network.interceptor.CloudflareInterceptor.intercept(CloudflareInterceptor.kt:46)
	... 34 common frames omitted
Caused by: java.nio.file.FileSystemAlreadyExistsException: null
	at jdk.zipfs/jdk.nio.zipfs.ZipFileSystemProvider.newFileSystem(ZipFileSystemProvider.java:104)
	at java.base/java.nio.file.FileSystems.newFileSystem(FileSystems.java:339)
	at java.base/java.nio.file.FileSystems.newFileSystem(FileSystems.java:288)
	at com.microsoft.playwright.impl.driver.jar.DriverJar.extractDriverToTempDir(DriverJar.java:116)
	at com.microsoft.playwright.impl.driver.jar.DriverJar.initialize(DriverJar.java:66)
	at com.microsoft.playwright.impl.driver.Driver.initialize(Driver.java:64)
	at com.microsoft.playwright.impl.driver.Driver.createAndInstall(Driver.java:101)
	... 41 common frames omitted
```

Seems like `Driver.createAndInstall` is calling `DriverJar.initialize` and which calls `extractDriverToTempDir` and it's trying to read from `driver/` in resources which wasn't bundled in my uber jar file. however I could find some files inside the `driver` dir directly inside the jar.

You can download a build artifact from here: https://github.com/Suwayomi/Tachidesk-Server-preview/releases/download/v0.6.6-r1165/Tachidesk-Server-v0.6.6-r1165.jar

Also this issue seems to be related to #1043 but I couldn't find anything about packaging PlayWright anywhere?

### PR DESCRIPTION
**Pull request by AriaMoradi**
Fixes #1138 

We verified this works with fat jars with our own.

Reference: https://github.com/Suwayomi/Tachidesk-Server/pull/462

### CODE REVIEW

File: driver-bundle/src/main/java/com/microsoft/playwright/impl/driver/jar/DriverJar.java [hunk 1]
```diff
@@ -106,14 +106,22 @@ private static boolean isExecutable(Path filePath) {
     return name.endsWith(".sh") || name.endsWith(".exe") || !name.contains(".");
   }
 
+private FileSystem initFileSystem(URI uri) throws IOException {
```
Comment (yury-s) [inline driver-bundle/src/main/java/com/microsoft/playwright/impl/driver/jar/DriverJar.java:109]: style nit: wrong indentation

**Review by yury-s — COMMENTED** [review]
The implementation looks good, can you add a test where the jar file system already exists and playwright would throw without the fix?

I believe you can do something like ./tools/test-cli-version and in the main test file create jar file system before calling playwright, something like

```java
    URI uri = new URI("jar:file:/home/user/.m2/repository/com/microsoft/playwright/driver-bundle/1.28.0/driver-bundle-1.28.0.jar!/driver/linux");
    FileSystem fs = FileSystems.newFileSystem(uri, Collections.emptyMap());
    assertNotNull(fs);
    try (Playwright playwright = Playwright.create()) {
      Browser browser = playwright.chromium().launch();
      Page page = browser.newPage();
    }

```

**Comment by AriaMoradi** [comment]
@microsoft-github-policy-service agree [company="Suwayomi Group"]

**Comment by AriaMoradi** [comment]
@microsoft-github-policy-service agree

**Comment by Syer10** [comment]
@microsoft-github-policy-service agree

File: .github/workflows/test_fatjar_cli.yml [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+          path: ~/.m2
+          key: ${{ runner.os }}-m2-${{ hashFiles('**/pom.xml') }}
+          restore-keys: ${{ runner.os }}-m2
+      - name: Download drivers
+        run: scripts/download_driver_for_all_platforms.sh
+      - name: Intall Playwright
+        run: mvn install -D skipTests --no-transfer-progress
+      - name: Test CLI version
+        shell: bash
+        run: tools/test-cli-farjar-version/test.sh
```
Comment (yury-s) [inline .github/workflows/test_fatjar_cli.yml:29]: Let's just merge this step into `.github/workflows/test_cli.yml` to avoid extra download & build of playwright on CI just for this test, we already do this for other cli tests.

File: tools/test-cli-fatjar-version/src/main/java/com/microsoft/playwright/testclifatjarversion/TestApp.java [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+import java.io.IOException;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.nio.file.FileSystem;
+import java.nio.file.FileSystems;
+import java.util.Collections;
+
+public class TestApp {
+  public static void main(String[] args) throws IOException, InterruptedException, URISyntaxException {
+    URI uri = new URI("jar:file:/home/user/.m2/repository/com/microsoft/playwright/driver-bundle/1.28.0/driver-bundle-1.28.0.jar!/driver/linux");
```
Comment (yury-s) [inline tools/test-cli-fatjar-version/src/main/java/com/microsoft/playwright/testclifatjarversion/TestApp.java:16]: Playwright version should be a parameter rather than hardcoded `1.28.0` value, so that it keeps working after bumping the version. [Same approach](https://github.com/microsoft/playwright-java/blob/81d31d24f82e5910c716529f0cdd3a5c5e61764d/tools/test-cli-version/src/main/java/com/microsoft/playwright/testcliversion/TestApp.java#L9) as in test cli version should probably work.

`/home/user/` will not work on other systems. I believe the easiest way to make it version/platform independent would be to extract [these lines](https://github.com/microsoft/playwright-java/blob/81d31d24f82e5910c716529f0cdd3a5c5e61764d/driver-bundle/src/main/java/com/microsoft/playwright/impl/driver/jar/DriverJar.java#L110-L113) into a method on DriverJar and then call it here (we don't have any promises about public API of DriverJar as it is under `.impl.` package so there are no issues with exposing such method just for the test. Or alternatively you can achieve the same via reflection.

File: tools/test-cli-fatjar-version/src/main/java/com/microsoft/playwright/testclifatjarversion/TestApp.java [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+    FileSystem fs = FileSystems.newFileSystem(uri, Collections.emptyMap());
+    assert (fs != null);
```
Comment (yury-s) [inline tools/test-cli-fatjar-version/src/main/java/com/microsoft/playwright/testclifatjarversion/TestApp.java:18]: Let's throw a RuntimeError instead this way it won't depend on whether assertions are enabled or not (I believe they are not enabled in the current CI config).

File: tools/test-cli-fatjar-version/src/main/java/com/microsoft/playwright/testclifatjarversion/TestApp.java (hunk 3 above)
Comment (Syer10) [reply tools/test-cli-fatjar-version/src/main/java/com/microsoft/playwright/testclifatjarversion/TestApp.java]: I am not sure if I am handling the last point correctly in the latest commit, which is why I split it from the other commit handling the review comments.

File: .github/workflows/test_cli.yml [hunk 5]
```diff
@@ -29,3 +29,6 @@ jobs:
       - name: Test CLI version
         shell: bash
         run: tools/test-cli-version/test.sh
+      - name: Test CLI Fatjar
+        shell: bash
+        run: tools/test-cli-farjar/test.sh
```
Comment (yury-s) [inline .github/workflows/test_cli.yml:34]:
```suggestion
        run: tools/test-cli-fatjar/test.sh
```

File: driver-bundle/src/main/java/com/microsoft/playwright/impl/driver/jar/DriverJar.java [hunk 6]
```diff
@@ -106,14 +106,25 @@ private static boolean isExecutable(Path filePath) {
...
-  void extractDriverToTempDir() throws URISyntaxException, IOException {
+  private FileSystem initFileSystem(URI uri) throws IOException {
+    try {
+      return FileSystems.newFileSystem(uri, Collections.emptyMap());
+    } catch (FileSystemAlreadyExistsException e) {
+      return null;
+    }
+  }
+
+  URI getPlatformDriver() throws URISyntaxException {
```
Comment (yury-s) [inline driver-bundle/src/main/java/com/microsoft/playwright/impl/driver/jar/DriverJar.java:117]: does it work without `public` modifier?

File: driver-bundle/src/main/java/com/microsoft/playwright/impl/driver/jar/DriverJar.java [hunk 7]
```diff
@@ -140,7 +151,7 @@ void extractDriverToTempDir() throws URISyntaxException, IOException {
           }
           toPath.toFile().deleteOnExit();
         } catch (IOException e) {
-          throw new RuntimeException("Failed to extract driver from " + uri + ", full uri: " + originalUri, e);
```
Comment (yury-s) [inline driver-bundle/src/main/java/com/microsoft/playwright/impl/driver/jar/DriverJar.java:143]: nit: having originalUri in the exception was useful for debugging when the jar was backed into some other jar (see https://github.com/microsoft/playwright-java/issues/306#issuecomment-797355207). We can live without it but maybe we can preserve it by using something like getDriverResourceURI in the test instead of getPlatformDriver and calling maybeExtractNestedJar only inside extractDriverToTempDir (this is for an edge case scenarios anyway and shouldn't affect your test).

**Review by yury-s — APPROVED** [review]

**Review by yury-s — APPROVED** [review]

File: tools/test-cli-fatjar/test.sh [hunk 8]
```diff
@@ -0,0 +1,13 @@
+#!/bin/bash
+
+set -e
+set +x
+
+trap "cd $(pwd -P)" EXIT
+cd "$(dirname $0)"
+
+echo "Running CLI..."
+mvn exec:java -e -Dexec.mainClass=com.microsoft.playwright.CLI -Dexec.args="install"
```
[the remark this answers is not in the document]
Comment (Syer10) [reply tools/test-cli-fatjar/test.sh:10]: The last actions run failed with this error
```kotlin
╔══════════════════════════════════════════════════════════════════════════════════════════╗
║ Looks like Playwright was just installed or updated.                                     ║
║ Please run the following command to download new browsers:                               ║
║                                                                                          ║
║     mvn exec:java -e -Dexec.mainClass=com.microsoft.playwright.CLI -Dexec.args="install" ║
║                                                                                          ║
║ <3 Playwright Team                                                                       ║
╚══════════════════════════════════════════════════════════════════════════════════════════╝
```

**Comment by Syer10** [comment]
Looks like the Test CLI action ran successfully this time around 🎉

File: tools/test-cli-fatjar/test.sh (hunk 8 above)
[the remark this answers is not in the document]
Comment (yury-s) [reply tools/test-cli-fatjar/test.sh:10]: Oh, I know what's going on, in your TestApp call Driver.createAndInstall instead of Driver.ensureDriverInstalled so that it doesn't change Driver.instance and when you call Playwright.create it will actually try to do all initialization.

[the remark this answers is not in the document]
Comment (Syer10) [reply tools/test-cli-fatjar/test.sh]: I see what happened, I used false as the downloadBrowsers prarmeter in Driver.ensureDriverInstalled, should be fixed with the latest commit.

[the remark this answers is not in the document]
Comment (Syer10) [reply tools/test-cli-fatjar/test.sh]: I use `Driver.ensureDriverInstalled` instead of `Driver.createAndInstall` so that playwright will use the same driver.

File: tools/test-cli-fatjar/src/main/java/com/microsoft/playwright/testclifatjar/TestApp.java [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+import java.io.IOException;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.nio.file.FileSystem;
+import java.nio.file.FileSystems;
+import java.util.Collections;
+
+public class TestApp {
+  public static void main(String[] args) throws IOException, URISyntaxException {
+    DriverJar driver = (DriverJar) Driver.ensureDriverInstalled(Collections.emptyMap(), true);
```
Comment (yury-s) [inline tools/test-cli-fatjar/src/main/java/com/microsoft/playwright/testclifatjar/TestApp.java:18]: If you do this, subsequent call Playwright.create() below will not actually trigger  DriverJar.initFileSystem second time (when a filesystem already exists). Just make DriverJar.getDriverResourceURI a static method and call it here without calling Driver.ensureDriverInstalled, it should work fine.

Comment (Syer10) [reply tools/test-cli-fatjar/src/main/java/com/microsoft/playwright/testclifatjar/TestApp.java]: Done 👍 
I wasn't able to do that before since `maybeExtractNestedJar` cannot be static

**Comment by AriaMoradi** [comment]
Is this ready to merge?

**Comment by yury-s** [comment]
Yes, it's ready to be merged, you can do that or I can merge later once the tests complete. Thanks for addressing all the comments!

**Review by yury-s — APPROVED** [review]
